File: 2. Structural Design Patterns/Bridge/c++/BridgeExample.cpp
#include <iostream>
using namespace std;

// Implementation: Device
class Device {
public:
    virtual void turnOn() = 0;
    virtual void turnOff() = 0;
    virtual ~Device() {}
};

class TV : public Device {
public:
    void turnOn() override {
        cout << "Turning on the TV\n";
    }
    void turnOff() override {
        cout << "Turning off the TV\n";
    }
};

class Radio : public Device {
public:
    void turnOn() override {
        cout << "Turning on the Radio\n";
    }
    void turnOff() override {
        cout << "Turning off the Radio\n";
    }
};

// Abstraction: RemoteControl
class RemoteControl {
protected:
    Device* device;
public:
    RemoteControl(Device* d) : device(d) {}
    virtual void on() {
        device->turnOn();
    }
    virtual void off() {
        device->turnOff();
    }
    virtual ~RemoteControl() {}
};

// Refined Abstraction: AdvancedRemote
class AdvancedRemote : public RemoteControl {
public:
    AdvancedRemote(Device* d) : RemoteControl(d) {}

    void mute() {
        cout << "Muting the device\n";
    }
};

int main() {
    TV tv;
    Radio radio;

    RemoteControl* remote1 = new RemoteControl(&tv);
    remote1->on();
    remote1->off();

    AdvancedRemote* remote2 = new AdvancedRemote(&radio);
    remote2->on();
    remote2->mute();
    remote2->off();

    delete remote1;
    delete remote2;

    return 0;
}

File: 2. Structural Design Patterns/Bridge/Python/BridgeExample.py
# Bridge Pattern Example in Python
# Remote control for different devices

from abc import ABC, abstractmethod

class Device(ABC):
    @abstractmethod
    def turn_on(self):
        pass
    
    @abstractmethod
    def turn_off(self):
        pass
    
    @abstractmethod
    def set_volume(self, volume):
        pass

class TV(Device):
    def __init__(self):
        self.volume = 50
        self.on = False
    
    def turn_on(self):
        self.on = True
        return "TV is ON"
    
    def turn_off(self):
        self.on = False
        return "TV is OFF"
    
    def set_volume(self, volume):
        self.volume = volume
        return f"TV volume set to {volume}"

class Radio(Device):
    def __init__(self):
        self.volume = 30
        self.on = False
    
    def turn_on(self):
        self.on = True
        return "Radio is ON"
    
    def turn_off(self):
        self.on = False
        return "Radio is OFF"
    
    def set_volume(self, volume):
        self.volume = volume
        return f"Radio volume set to {volume}"

class RemoteControl:
    def __init__(self, device):
        self.device = device
    
    def power(self):
        return self.device.turn_on()
    
    def volume_up(self):
        return self.device.set_volume(self.device.volume + 10)
    
    def volume_down(self):
        return self.device.set_volume(self.device.volume - 10)

class AdvancedRemote(RemoteControl):
    def mute(self):
        return self.device.set_volume(0)

def main():
    tv = TV()
    tv_remote = RemoteControl(tv)
    print(tv_remote.power())
    print(tv_remote.volume_up())
    
    radio = Radio()
    radio_remote = AdvancedRemote(radio)
    print(radio_remote.power())
    print(radio_remote.mute())

if __name__ == "__main__":
    main()
